File: cpp/876. Middle of the Linked List.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;
typedef vector<int> vi;
typedef vector<vector<int>> vvi;
typedef pair<int, int> pi;


struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
    ListNode* middleNode(ListNode* head) {
        ListNode *fast = head, *slow = head;       
        while(fast && fast->next) {
            slow = slow->next;
            fast = fast->next->next;
        }
        return slow;
    }
};

int main() {
    ios::sync_with_stdio(0); 
    cin.tie(0);

    
}

File: python/876-middle-of-the-linked-list.py
from linklist_helper import ListNode, gen_linklist, display_linklist

'''
Two pointer slow, fast
When fast go to last, the slow is the middle point
https://www.youtube.com/watch?v=wmpivqMlClI
'''
class Solution:
    '''
        [1,2,3,4,5] -> 3
        [1,2,3,4,5,6] -> 4
    '''
    def middleNode(self, head: ListNode) -> ListNode:
        "Time: O(n), Space: O(1)"
        slow, fast = head, head
        while fast is not None and fast.next is not None:
            slow = slow.next
            fast = fast.next.next
        return slow

l = [1,2,3,4,5]
l = [1,2,3,4,5,6]
head = gen_linklist(l)
result = Solution().middleNode(head)
print(result)